### PR TITLE
Fix auto-confirm on Xiaomi Mi8 (fixes #39)

### LINKED ISSUE #39: Problem with auto-entering backup pw and confirm
Problem / Reproducer
Hi there,
thanks for this wonderful script. Since I've decided to stay unrooted for the moment, I'm using it to backup my apps and data. I've got a problem here, as the auto-confirm config option doesn't work. To reproduce this:
- Use the sample configuration. (Auto confirm is on there)
- Define the var "BACKUP_PASSWORD".
- Run main script with subdir name "_mi8". cd to "_mi8" folder.
- Run "./userbackup" script.
- It shows "Now unlocking your device" and the first package name to backup. The Android device pops up the confirm desktop backup screen from Android and requires me to enter the backup password as I'd previously defined in developer settings.
- Not touching the phone while this is happening:
adb shell input [password] is sent by the desktop and the password appears in the first (from the two password) text fields on the Android screen. 
At the same time: Android keyboard pops up.
PROBLEM: The backup gets stuck here, I have to manually hit the "Backup now" button on the bottom right corner.
EXPECTED: The "Backup now" button is automatically hit by the correct keyevent the desktop has to send to the Android device.

I also found a fix for the problem. Will submit a PR.

Device
Xiaomi Mi8

Android OS
MIUI 10 from Xiaomi, Android 8.1.0 Oreo based.

Desktop OS
Windows 10 Pro x64 using GIT bash based on MinGW.

Maybe related to issue:
#36 "Confirmation on every backup / restore"

### PR DESCRIPTION
**Pull request by ghost**
Purpose
Fix the keys sent to the android device needed to hit the confirm backup button on the bottom right screen corner.

Testing
Verified working on Android 8.1.0, Xiaomi MIUI 10 on Xiaomi Mi8 device.

Sidenote
The sleep's were removed as the backup worked perfectly without them. I also confirmed that the files backed up were "greater than 0 bytes".

### CODE REVIEW

**Comment by IzzySoft** [comment]
Thanks, that might fix it for Xiami – but it might also break it for all devices it was working with before, as you removed the existing keyevents. Have you tested this with devices other than Xiaomi? I will test it on my devices as soon as I find some time, but cannot promise when that will be (hopefully this month, though – but this week is unlikely). If it breaks, we could consider making the two events (22/23 in the original code, 61/66 in yours) configurable per-device, with a default set. But:

Removing the sleeps because it worked on *your* devices is no good idea. I have a device where the sleep was not needed, and another were it doesn't work without (and got several reports for each case from others). Guess we can live with the 1s delay better than with the feature not working for some, agreed?

**Comment by ghost** [comment]
Thanks for your feedback on this.
( Guess we can live with the 1s delay better than with the feature not working for some, agreed? )
Sure. I didn't try it on another device yet. I suspect the difficulty will come from different android versions which handle the confirmation dialog differently.

**Comment by IzzySoft** [comment]
Indeed. And the number of devices we have to check with is rather limited – so we cannot say for sure if any of those needing the sleep are left or not. I'd rather like not to break stuff for others.

As said, as soon as I find some time, I'll check your changes with the devices I have available.

File: lib/scriptgen.lib [hunk 1]
```diff
@@ -209,11 +209,11 @@ getUserAppBackup() {
   if [[ "${AUTO_CONFIRM}" -eq 1 ]]; then
       echo "function autoConfirm() {" >> "$backupscript"
       if [[ ! -z "${BACKUP_PASSWORD}" ]]; then
-          echo -e "    adb ${ADBOPTS} shell input text \"${BACKUP_PASSWORD}\"" >> "$backupscript"
+          echo -e "    adb ${ADBOPTS} shell input text \"${BACKUP_PASSWORD}\"; sleep 0.5" >> "$backupscript"
       fi
-      echo -e "    adb ${ADBOPTS} shell input keyevent 61" >> "$backupscript"
-      echo -e "    adb ${ADBOPTS} shell input keyevent 61" >> "$backupscript"
-      echo -e "    adb ${ADBOPTS} shell input keyevent 61" >> "$backupscript"
+      echo -e "    adb ${ADBOPTS} shell input keyevent 61; sleep 0.3" >> "$backupscript"
```
Comment (IzzySoft) [inline lib/scriptgen.lib:214]: that breaks it for me on MM (LG P880 with LOS 13, Wilefox Swift with CyanogenOS 13) as well as N (BQ Aquaris X5 Plus with LOS 14.1) – stopped trying more devices after that. On all of them, the "old" code worked. With your code, even though no BACKUP_PASSWORD is set, it activates the password field, then puts a dot in, and that's it – no more action, it just hangs there.

Further: why is this needed to be repeated multiple times? Keycode 61 is `KEYCODE_TAB` (see [this list of keycodes](https://gist.github.com/ctrl-freak/9abb5aea0d89d7bd9df6a3d0ac08b73c)). So this means you need 3x tab, then Enter (Keycode 66=`KEYCODE_ENTER`) – while on the devices I just tested with (the original code of this) it is 1x right (22: `KEYCODE_DPAD_RIGHT`), then a different Enter (23: `KEYCODE_DPAD_CENTER`). And no, that device has no DPAD.

That means we might need ***device-specific*** keycode ***sequences.*** Just fixing it for your Xiaomi breaks it for several others, as I was afraid of.

Examples:

``` bash
    BACKUP_CONFIRM_SEQUENCE=(22 23)       # my devices
    BACKUP_CONFIRM_SEQUENCE=(61 61 61 66) # your Xiaomi
    for keyevent in ${BACKUP_CONFIRM_SEQUENCE[@]}; do
      adb ${ADBOPTS} shell input keyevent ${keyevent}; sleep 0.5
    done
```

Note I also increased `sleep` again. You reduced it from 1s to 0.3s. Not all devices are that fast (consider low-end ones). 0.5s should be a good compromise (we still can increase if there are complaints coming in, or make it a setting as well).

**Comment by ghost** [comment]
Thanks for taking the time to test it. So lets agree on 0.5 secs of sleep always and a device specific sequence. I manually tested the 22/23 keycodes and they had no effect on xiaomi miui 10 .

**Comment by IzzySoft** [comment]
OK, just give me a little. And we keep this PR open until a solution is found (reminder for me). As I wrote, "I have a plan" already. Just takes a little for clean implementation and correct documentation.

**Comment by IzzySoft** [comment]
OK, can you please test (with HEAD – relevant changes were committed with c84081e)? Variables you will want to set/adjust in your config:

* `MK_AUTOCONFIRM_SEQUENCE=(61 61 61 66)`
* `MK_AUTOCONFIRM_DELAY=0.3` (your preferred value)

**Comment by IzzySoft** [comment]
Cool, so it works for you as well! Then I should commit the documentation (wiki) now. Thanks for closing!

**Comment by ghost** [comment]
I posted in the issue, so we can go on making the best out of the different key seq and dialog behaviours...

**Pull request closed without merging**